Add ECCBUS attribute and use it for gateway_mfr

radicale opened the pull request:
**Issue**
Summary: The library currently uses a Balena environment variable to determine the device type and choose ECC device for gateway_mfr in the case of RockPI.  It also does not provide a mechanism to support other hardware platforms where the ECC device is not /dev/i2c-1.  Besides this method making the system depend on Balena, it is also incoherent with the objective of defining hardware specific attributes in the library itself.

**How**
- Added a new attribute "ECCBUS" and modified run_gateway_mfr() to use this attribute instead of is_rockpi().
- Modified the hardware definitions to include the new attribute.

**References**
#69 

**Checklist**

- [x] Tests added
- [x] Cleaned up commit history (rebase!)
- [ ] Documentation added
- [x] Thought about variable and method names

marvinmarnold commented on hm_pyhelper/hardware_definitions.py:
```diff
@@ -34,6 +34,7 @@ def is_raspberry_pi():
         'FRIENDLY': 'Nebra Indoor Hotspot Gen 1',
         'APPNAME': 'Indoor',
         'SPIBUS': 'spidev1.2',
+        'ECCBUS': 'i2c-1',
```
I think `ECC_I2C_BUS` would be better. Our current keys like `SPIBUS` are not as clear as they could be.

radicale replied:
Yup, just went along with the `SPIBUS` naming convention here.  Also used only a device name for coherence.

marvinmarnold commented on hm_pyhelper/miner_param.py:
```diff
@@ -26,9 +27,12 @@ def run_gateway_mfr(args):
-    if is_rockpi():
-        extra_args = ['--path', '/dev/i2c-7']
+    try:
+        extra_args = ['--path', "/dev/" +
```
Can you split this up into different lines for legibility? Like:

```
ecc_path = 'dev' + get_variant_attribute(os.getenv('VARIANT'), 'ECCBUS')
extra_args = ...
```

marvinmarnold commented on hm_pyhelper/miner_param.py:
```diff
@@ -26,9 +27,12 @@ def run_gateway_mfr(args):
+                      get_variant_attribute(os.getenv('VARIANT'), 'ECCBUS')]
         command.extend(extra_args)
+    except (UnknownVariantException, UnknownVariantAttributeException) as e:
+        LOGGER.exception(e)
```
Should not swallow the exception. Also recommend a clearer logging message: "Unable to call gateway_mfr because ECC path could not be determined"

radicale replied:
Actually, if those errors occur, it would just call gateway_mfr without the path parameter (hence default to i2c-1).  This was my intended behavior.  I suppose I could be more explicit with the message and also set `extra_args` to `None`.

marvinmarnold reviewed (dismissed):
Hey @radicale thanks for opening a PR. In addition to the other comments, this PR is missing tests: https://github.com/NebraLtd/hm-pyhelper/blob/40bdda8be01da7e0988a825ff091fddcea06f444/hm_pyhelper/tests/test_hardware_definitions.py#L12

I'd also like to point out that we will be reviewing and prioritizing community PRs on a case-by-case basis. Thanks for making a PR that addresses one of our open issues!

shawaj commented:
One other thought - more for the future than necessarily for this PR - but with 5G we will use TPM2.0 and we are also investigating other ECC chips from Maxim Integrated as well as ARM TrustZone.

Perhaps we should look into abstracting away even further or having another attribute of `KEY_STORAGE` or similar with options of `ECC`, `TRUSTZONE`, `TPM2`, `MAXIM` or something like that.

Not sure what you think.

shawaj commented:
Also would be good to add info about the new key to the README

marvinmarnold commented:
> Perhaps we should look into abstracting away even further or having another attribute of `KEY_STORAGE_METHOD` or similar with options of `ATECC`, `TRUSTZONE`, `TPM2`, `MAXIM` or something like that. Then instead of `ECC_I2C_BUS` as @marvinmarnold mentioned we could use `KEY_STORAGE_BUS` and that could then cover other methods of connection as well. And as part of that we could use the full path like `/dev/i2c-1` in case it is no longer on `/dev/` root for example

@shawaj I like that a lot. `KEY_STORAGE_BUS` plus full path `/dev/i2c-1`.

radicale commented:
I've implemented the requested changes:
- KEY_STORAGE_BUS instead of ECCBUS
- Full device path instead of assuming initial "/dev/"
- The attribute remains optional, but added a warning level log entry if it fails lookup
- Added tests on the gateway_mfr arguments when the attribute is present and not

shawaj commented:
Looks good to me @radicale - thanks!

Can you bump version number in [setup.py](https://github.com/NebraLtd/hm-pyhelper/blob/40bdda8be01da7e0988a825ff091fddcea06f444/setup.py#L6) as well?

radicale commented:
Had to modify the tests as I as using an attribute only present in python => 3.8, but the workflow is using 3.7.  Renamed tests and added another test case for invalid variant.  Cleaned up commit history, and bumped version.

shawaj commented:
Thanks very much @radicale - very much appreciated 👌

marvinmarnold commented:
Yea thanks @radicale! Appreciate your changes. Happy Holidays.

shawaj commented:
@radicale noticed you are working on balena support for the pisces p100.

Feel free to add PRs for other things such as the new hardware variant etc etc and any other modifications (was thinking we could use the key storage bus parameter in the docker compose workflow also to automatically pull in the i2c bus as well)

We don't have a Pisces miner here at the moment so haven't been able to add it.

shawaj commented:
@radicale noticed you are working on balena support for the pisces p100.

Feel free to add PRs for other things such as the new hardware variant etc etc and any other modifications (was thinking we could use the key storage bus parameter in the docker compose workflow also to automatically pull in the i2c bus as well)

We don't have a Pisces miner here at the moment so haven't been able to add it. But would be great to have the support👌

radicale commented:
@shawaj I certainly will do.  I've really just cobbled things together for now, but as I polish things up, I will make more pull requests.

The docker device parameters are an interesting pickle to solve while being consistent with hm-pyhelper.  As you mention, it could be done through a workflow.  My thoughts were to use variable substitution in the docker compose file as well as a workflow.  This way, automatically built versions will have the correct device path, but the source file would still work on it own if you set the environment variable in the host OS.  Also adds a convenient target in the file for the workflow to substitute.

radicale commented:
So I read that balena deploy does not support variable substitution.  It's an open [issue](https://github.com/balena-io-modules/balena-compose-parse/issues/50).  Still perhaps worthwhile to use the syntax just to have it as a regex target for a wf, but it may also be a source of confusion for those that expect it to work with balena as-is.

kashifpk commented:
@radicale @shawaj - I believe that work in https://github.com/nebraltd/helium-miner-software/issues/272 should help with this. We're using jinja2 templates for variable substitution and some other stuff. Balena not supporting variable substitutions was one of the reasons we started work on 272.

radicale commented:
@kashifpk that fits the bill exactly.  Running the file through docker-compose beforehand in the wf should allow to use it for the device specific parameters too.  Can even use default values like so:
```
    devices:
      - "${HM_KEY_STORAGE_BUS:-/dev/i2c-1}:${HM_KEY_STORAGE_BUS:-/dev/i2c-1}"

radicale commented:
I hadn't read through it all, too bad for balena not liking the docker-compose output.  Not sure why it shouldn't though...
